### ISSUE TITLE
 Csla.Xaml.PropertyInfo & XamarinForm

### ISSUE THREAD

**Issue by sylum** [MarimerLLC/cslaforum#443]
Hi Folks,

I'm trying to use the PropertyInfo control in a XamarinForm app, and I can't get the binding to work correctly.  I want to bind another control's appearance to the IsValid meta data for a BO property that the PropertyInfo control exposes. 

Unfortunatly, I think I not doing it right , because when I tried the following code, i got the exception : **Method not found: 'Void Csla.Xaml.PropertyInfo.set_Path(System.String)'.**

```
<DatePicker  x:Name="productLastControlDate" Date="{Binding Model.LastControlDate}" />
<csla:PropertyInfo x:Name="productLastControlDatePropertyInfo"  BindingContext="{Binding .}" Path="Model.LastControlDate" />
<Label BindingContext="{x:Reference productLastControlDatePropertyInfo}"  Text="{Binding IsValid}" />
```
 
if anyone can advise me  about use properly the PropertyInfo control. 

thx in advance for your help,

Cedric

**Comment by rockfordlhotka** [comment]
I'm running off to take care of some things, but wanted to do at least a quick answer first.

Xamarin data binding is very different from WPF/SL/WRT/UWP, in that each UI control binds to _exactly one source_. Normal XAML binding allows each property to bind to a different source, but in Xamarin the entire control (all its properties) bind to a single source.

So the `PropertyInfo` "control" for Xamarin is quite different from the other XAML flavors. It is designed as an aggregator instead of as a separate source.

In other words, bind your UI control to the PropertyInfo (because the UI control can only have one source), and then bind the properties of your UI control to the properties of the PropertyInfo object.

And bind the PropertyInfo object to your domain model.

You'll notice that the Xamarin PropertyInfo object has a `Value` property, which is how you get/set the property of the underlying business object.

**Comment by sylum** [comment]
Hi Rocky,

Many thanks for the explication on the PropertyInfo . That helped me understand how it's work.

But I think, I still missing something about the binding on the PropertyInfo.
Actually, when I tried to  bind the PropertyInfo object to my domain model, I got an exception.

`<csla:PropertyInfo x:Name="productLastControlDatePropertyInfo"  BindingContext="{Binding Model}" Path="LabelCode" />`

I got the following error  :

> An exception of type 'System.MissingMethodException' occurred in RFitMobile.XamarinForms.dll but was not handled in the user code
> Additional information: Method not found: 'Void Csla.Xaml.PropertyInfo.set_Path (System.String)'.

I'm confident about access to the binding domain model cause the TextCell display well the data

`<TextCell BindingContext="{Binding Model}" Detail="{Binding LabelCode}"></TextCell>`

I will keep going the investigation.

**Comment by sylum** [comment]
Hi Folks,

I'm still struggling with **Csla.Xaml.PropertyInfo** on **XamarinForm**.

So, I got the ProjectTracker solution and  added a new UWP projet  "ProjectTracker.Ui.XFUWP" which reference the XamarinForm projet "XamarinFormsUi".

In the "ProjectEdit.xaml" I have uncommented the ligne below the check the **csla:PropertyInfo** behaviour 
`<csla:PropertyInfo x:Name="NameInfo" Path="Model.Name"/>`

Next,  I run the XamarinForm projet "XamarinFormsUi" and on the ProjectEdit screen => I have got an exception :

> Method not found: 'Void Csla.Xaml.PropertyInfo.set_Path (System.String)

By the way, the **Csla.Xaml.PropertyInfo** used into the projet "ProjectTracker.Ui.UWP" work well.

Finally, I really dont know what I'm doing wrong. 
Below, the link to the demo (including ProjectTracker.Ui.XFUWP)  with github or Ondrive.

[https://github.com/sylum/ProjectTracker](https://github.com/sylum/ProjectTracker)

[https://1drv.ms/u/s!AkXVxJpqEXeFgYczBVjsaYcqQpGKkg](https://1drv.ms/u/s!AkXVxJpqEXeFgYczBVjsaYcqQpGKkg)

Thanks in advance for your help,

Best regards,

Cedric

**Comment by Bowman74** [comment]
I'll sett if I can't get the Project Tracker project in working order this evening, perhaps that will help one way or the other.

**Comment by sylum** [comment]
This is great, many thanks for your help.

**Comment by Bowman74** [comment]
I have submitted a pull request to your project tracker repository that fixes the property info in your example. 

Let me know if you have any problems.

**Comment by Bowman74** [comment]
@sylum let me know if that resolved your question. Happy to help if you have more problems. Thanks.

**Comment by sylum** [comment]
Hi Kevin

Many thanks for your submitted.

But even with you code (we tried it on 3 differents PC) we still get the Exception  '_System.MissingMethodException_' raised by the method **InitializeComponent()** from **ProjectEdit**

> Method not found: 'Void Csla.Xaml.PropertyInfo.set_Path (System.String)

![devenv_2017-10-20_12-47-18](https://user-images.githubusercontent.com/15035848/31821214-a5c8120a-b5a4-11e7-82c2-704cbdd5430b.png)

I'm lost cause, I sure your code is running properly but dont understand what is happening.
- I have also tried by changing XamarinForm 2.3.4.247 to 2.2.0.31 => Same issue.
- I have also tried by switching the CultureInfo to 'US' => Same issue
- Each time, I'm clear all projects and delete files into 'bin' and 'obj' folder.

I keep looking on it.

**Comment by Bowman74** [comment]
OK, I got it working for iOS and Android. I didn't really look at the UWP app. If that's where your problem lays I'll take a look there too.

thanks,

Kevin

**Comment by sylum** [comment]
You right, the issue come from UWP apps. It would be great to have your feed back on it.

Thanks,

Cedric

**Comment by sylum** [comment]
Hi Kevin,
I have test the **Csla.Xaml.PropertyInfo** under **XamarinForm Android** projet and works well.
So I think there may be an issue when running on the **XamarinForm UWP**.
Have you tried it on **XamarinForm UWP** ?
Thanks,
Cedric

**Comment by Bowman74** [comment]
@sylum I got it. Yes, I think you have found a bug after looking through the sourcecode.

@rockfordlhotka Here's the deal on what I think is going on here. As you pointed out earlier in this thread, the UWP binding and Xamarin Forms binding are very different. The problem is Xamarin Forms works on UWP. So when @sylum gets the nuget package for his UWP project he's getting the one compiled for UWP bindings but is trying to use it in Xamarin Forms Xaml and Xamarin Forms doesn't have an idea how to use it, hilarity ensues... They are completely different, derive from different base classes, it's all bad.

I think the base problem here (pardon the pun) was the idea that the base class for this item a project needs is determined by the platform when it is really determined the type of XAML in play. I believe that the Forms CSLA Xaml needs to be a different Nuget package than the one for UWP Xaml or have two different classes. I like the first solution best, different nuget packages ,but whatever works. 

@sylum What you can probably do for a temporary workaround is edit your UWP project file to point at the iOS version of CSLA Xaml. I don't think there is anything iOS specific in there (I hope). Change that line to be something like:

..\..\packages\CSLA-XamarinForms.[VERSION YOU ARE USING]\lib\Xamarin.iOS\Csla.Xaml.dll

This workaround will only work if you have no native UWP Views that need CSLA Xaml and are only creating Xamarin Forms views.

**Comment by sylum** [comment]
Hi Kevin,

Brilliant, your solution of using the **Csla.Xaml.dll** from **Xamarin.iOS** works nicely.
As temporary workaround, I have made a batch process to copie the dll file into the bin folder after compilation. 

Many thanks to you for your help.

Cedric

**Comment by rockfordlhotka** [comment]
@Bowman74 I suspect I'll need to change the control name or namespace. NuGet doesn't (to my knowledge) have any way of differentiating based on the tech you are using, only based on the platform your project targets.

So I'll need to figure out a way to get both implementations of "PropertyInfo" into all the Csla.Xaml assemblies - which means a different name or namespace.

I suspect a different name is the way to go - thus avoiding confusion.

Any suggestions on a different name?

**Comment by Bowman74** [comment]
@rockfordlhotka Yes, either two nuget packages with different namespaces or two different classes in two namespaces in the same package. I don't know Csla.Forms.Xaml or something. Breaking change no matter how you go.

**Comment by Bowman74** [comment]
@sylum My pleasure.
